### PR TITLE
Randomize mac address on release if restoring original one fails

### PR DESCRIPTION
**Pull request by booxter**
Some NICs / drivers and some MAC addresses may fail to restore (for
example, because a driver / vendor doesn't support an all-zeroes MAC
address that at the same time is the original MAC address assigned to
VFs on boot). In this case, it may be prudent to try to reset the
address to something else.

In this patch, we randomize the address instead of assigning a single
hardcoded one to address the issue where multiple VFs that originally
failed to restore the MAC address and were assigned the same address
on release are then re-allocated to new pods with the hardcoded
address left in place, which then result in duplicate addresses and
potential connectivity issues.

Signed-off-by: Ihar Hrachyshka <ihrachys@redhat.com>

### CODE REVIEW

**Comment by booxter** [comment]
Testing it here: https://github.com/kubevirt/kubevirt/pull/3098

**Comment by booxter** [comment]
Please hold it off, the relevant kubevirt job failed once so I need to understand why (whether the patch works as I expected).

**Comment by zshi-redhat** [comment]
@booxter Should the VF be assigned with a random effective MAC address already during VF provisioning?

**Comment by booxter** [comment]
@zshi-redhat why? what would make it get a random MAC address? Unless you use some additional mechanism to override original MAC addresses, they may be not so random. (Like, some drivers leave all VFs with all-zeroes MACs on boot).

**Comment by zshi-redhat** [comment]
> @zshi-redhat why? what would make it get a random MAC address? Unless you use some additional mechanism to override original MAC addresses, they may be not so random. (Like, some drivers leave all VFs with all-zeroes MACs on boot).

@booxter So even for effective MAC address, the original value when first provisioned is all-zero? I was under the impression that it's admin MAC address that gets all-zero when first created.

File: pkg/sriov/sriov.go [hunk 1]
```diff
@@ -245,8 +259,14 @@ func (s *sriovManager) ReleaseVF(conf *sriovtypes.NetConf, podifName string, cid
 				return fmt.Errorf("failed to parse original effective MAC address %s: %v", conf.EffectiveMAC, err)
 			}
 
-			if err = s.nLink.LinkSetHardwareAddr(linkObj, hwaddr); err != nil {
-				return fmt.Errorf("failed to restore original effective netlink MAC address %s: %v", hwaddr, err)
+			if originalErr := s.nLink.LinkSetHardwareAddr(linkObj, hwaddr); originalErr != nil {
```
Comment (adrianchiris) [inline pkg/sriov/sriov.go]: Just a thing that pops to mind,
if `conf.EffectiveMAC` is all zeroes then set  a random MAC, else restore to the saved value.
then we would not trigger netlink commands twice.

if  LinkSetHardwareAddr  fails  from other reasons, there may be an issue with the system,  so the second call will also fail.

i dont think we need to worry about collisions

File: pkg/sriov/sriov.go [hunk 2]
```diff
@@ -207,6 +208,19 @@ func (s *sriovManager) SetupVF(conf *sriovtypes.NetConf, podifName string, cid s
...
+// Generate a random mac for interface
+// Avoid MAC address starting with reserved value 0xFE (https://github.com/kubevirt/kubevirt/issues/1494)
+// NOTE: this function is borrowed (nearly) verbatim from kubevirt/kubevirt github repo
+func generateRandomMac() (net.HardwareAddr, error) {
+	prefix := []byte{0x02, 0x00, 0x00} // local unicast prefix
+	suffix := make([]byte, 3)
+	_, err := rand.Read(suffix)
+	if err != nil {
+		return nil, err
+	}
```
Comment (adrianchiris) [inline pkg/sriov/sriov.go]: To support a wider random values you could do something like:

```
func generateRandomMac() (net.HardwareAddr, error) {
	macBuf := make([]byte, 6)
	_, err := rand.Read(macBuf)
	if err != nil {
		return nil, err
	}
	macBuf[0] = macBuf[0] & 0xfe  // ensure link-local bit is set and mcast bit cleared
	return net.HardwareAddr(macBuf), nil
}
```

**Comment by zshi-redhat** [comment]
@booxter sorry for the late response on this PR. 

Just chatted with @ahalim-intel and we got a few questions below:
1) Does the issue happen on kernel interface or userspace(dpdk) interface? We thought kubevirt use vfio driver which is dpdk interface, but the change in this PR is about effective mac address. (sr-iov cni doesn't configure effective mac for dpdk interface)
2) Could you please share the exact NIC type? We are not sure if there is any device that will have initial effective mac address being all zero.

**Comment by martinkennelly** [comment]
Closing due to inactivity. Please re-open if you want to address this again.

**Pull request closed without merging**